File: server/app/models/article.py
from sqlalchemy import Column, Integer, String, Text, Boolean, ForeignKey, DateTime, Float, ARRAY, UniqueConstraint
from sqlalchemy.sql import func
from blog.extensions import db

class Article(db.Model):
    """Articles model for storing articles with vector embeddings."""
    __tablename__ = 'articles'

    id = Column(Integer, primary_key=True, autoincrement=True)
    image = Column(Text, nullable=True)
    slug = Column(String(255), unique=True, nullable=False)
    title = Column(String(255), nullable=False)
    content = Column(Text, nullable=False)
    author = Column(Integer, ForeignKey('users.id'), nullable=False)
    created_at = Column(DateTime, nullable=False, default=func.now()) 
    is_draft = Column(Boolean, nullable=False, default=True)
    embedding = Column(ARRAY(Float), nullable=True) 

    def __repr__(self):
        return f"<Article {self.title}>"

    def to_dict(self):
        return {
            'id': self.id,
            'image': self.image,
            'slug': self.slug,
            'title': self.title,
            'content': self.content,
            'author': self.author,
            'created_at': self.created_at.isoformat(),
            'is_draft': self.is_draft,
            'embedding': self.embedding
        }


class Tag(db.Model):
    """Tags model for storing unique article tags."""
    __tablename__ = 'tags'

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(255), unique=True, nullable=True)

    def __repr__(self):
        return f"<Tag {self.name}>"

    def to_dict(self):
        return {
            'id': self.id,
            'name': self.name
        }
        
        
class ArticleTag(db.Model):
    """ArticleTags junction model to link articles and tags."""
    __tablename__ = 'article_tags'

    article_id = Column(Integer, ForeignKey('articles.id'), nullable=False)
    tag_id = Column(Integer, ForeignKey('tags.id'), nullable=False)

    __table_args__ = (UniqueConstraint('article_id', 'tag_id', name='uix_article_tag'),)

    def __repr__(self):
        return f"<ArticleTag article_id={self.article_id}, tag_id={self.tag_id}>"

    def to_dict(self):
        return {
            'article_id': self.article_id,
            'tag_id': self.tag_id
        }

File: internal/models/article.go
package models

type Article struct {
	ID                       int64  `gorm:"primaryKey" json:"id"`
	Image                    string `gorm:"type:varchar(255)" json:"image"`
	Slug                     string `gorm:"type:varchar(255);uniqueIndex" json:"slug"`
	Title                    string `gorm:"type:varchar(255)" json:"title"`
	Content                  string `gorm:"type:text" json:"content"`
	Author                   int64  `gorm:"index" json:"author"`
	CreatedAt                int64  `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt                int64  `gorm:"autoUpdateTime" json:"updated_at"`
	IsDraft                  bool   `gorm:"default:true" json:"is_draft"`
	Embedding                []byte `gorm:"type:bytea" json:"-"`
	ImageGenerationRequestID string `gorm:"type:varchar(255)" json:"image_generation_request_id"`
	PublishedAt              *int64 `gorm:"index" json:"published_at,omitempty"`
	Tags                     []Tag  `gorm:"many2many:article_tags;" json:"-"`
}

type Tag struct {
	ID       int64     `gorm:"primaryKey" json:"id"`
	Name     string    `gorm:"type:varchar(255);uniqueIndex" json:"name"`
	Articles []Article `gorm:"many2many:article_tags;" json:"-"`
}

type ArticleTag struct {
	ArticleID int64   `gorm:"primaryKey" json:"article_id"`
	TagID     int64   `gorm:"primaryKey" json:"tag_id"`
	Article   Article `gorm:"foreignKey:ArticleID;constraint:OnDelete:CASCADE"`
	Tag       Tag     `gorm:"foreignKey:TagID;constraint:OnDelete:CASCADE"`
}
